src/webview/js/js.js

```diff
@@ -466,31 +493,48 @@ const scrollToPreserve = (msgId: number, prevBoundTop: number) => {
   window.scrollBy(0, newBoundRect.top - prevBoundTop);
 };
 
-const appendAuthToImages = auth => {
-  const imageTags = document.getElementsByTagName('img');
-  arrayFrom(imageTags).forEach(img => {
-    if (!img.src.startsWith(auth.realm)) {
-      return;
-    }
+/**
+ * Rewrite the source URLs of all relative-URL <img> tags beneath the specified
+ * parent element to have an absolute path referencing the Zulip realm.
+ *
+ * DEPRECATED: If no parent element is specified, we transform every <img> in
+ * the entire document.
+ */
+const appendAuthToImages: (auth: Auth, parent?: Element | Document) => void = (() => {
+  // List of routes which accept the API key appended as a GET parameter.
+  const inlineApiRoutes: string[] = ['/user_uploads/', '/thumbnail?', '/avatar/'];
+
+  return (auth: Auth, parent: Element | Document = document) => {
+    const imageTags = parent.getElementsByTagName('img');
+    arrayFrom(imageTags).forEach(img => {
+      // Get the raw `src` value from the DOM. (We can't easily use `img.src`,
+      // since it's absolutized by the browser.)
+      //
+      /* $FlowFixMe: upstream `getNamedItem` return type should be `Attr | null` */
+      const actualSrc: string | null = img.attributes.getNamedItem('src')?.value;
+
+      // Skip elements with non-relative sources.
+      if (actualSrc && !actualSrc.startsWith('/')) {
```

Comment:
This is a bit more specific than "relative", right? Like if the value were `../foo.png` or `foo.png` or `?x=y` or `#sec-1`, or in the other direction `//chat.example.com/user_uploads/foo.png`, those would be relative and not matched here.

Basically this assumes that any relative URL reference we'll see as the `src` of an `img` is one replacing the whole path, and no more or less. That assumption sounds quite plausible, and I think it's fine to make if we're explicit about it and we confirm the server code meets it. Mostly I just want to be sure to be explicit about it. :slightly_smiling_face:

Reply:
> On the other hand, this is inside the WebView, and URL objects have been around for a very long time. Rewritten to explicitly let the browser parse URLs rather than doing it by hand,

SGTM!